odxtools/commrelation.py

```diff
@@ -0,0 +1,126 @@
+
+        in_param_if_snref = None
+        if (in_param_if_snref_elem := et_element.find("IN-PARAM-IF-SNREF")) is not None:
+            in_param_if_snref = odxrequire(in_param_if_snref_elem.get("SHORT-NAME"))
+
+        if et_element.find("IN-PARAM-IF-SNPATHREF") is not None:
+            warnings.warn("SNPATHREFs are not supported by odxtools yet", OdxWarning, stacklevel=1)
+
+        out_param_if_snref = None
+        if (out_param_if_snref_elem := et_element.find("OUT-PARAM-IF-SNREF")) is not None:
+            out_param_if_snref = odxrequire(out_param_if_snref_elem.get("SHORT-NAME"))
+
+        if et_element.find("OUT-PARAM-IF-SNPATHREF") is not None:
+            warnings.warn("SNPATHREFs are not supported by odxtools yet", OdxWarning, stacklevel=1)
+
+        value_type_raw = None
+        if (value_type_str := et_element.get("VALUE-TYPE")) is not None:
+            try:
+                value_type_raw = CommRelationValueType(value_type_str)
+            except ValueError:
+                odxraise(f"Encountered unknown comm relation value type '{value_type_str}'")
+
+        return CommRelation(
+            description=description,
+            relation_type=relation_type,
+            diag_comm_ref=diag_comm_ref,
+            diag_comm_snref=diag_comm_snref,
+            in_param_if_snref=in_param_if_snref,
+            out_param_if_snref=out_param_if_snref,
+            value_type_raw=value_type_raw)
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        return {}
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        if self.diag_comm_ref is not None:
+            self._diag_comm = odxlinks.resolve(self.diag_comm_ref, DiagComm)
+
+    def _resolve_snrefs(self, context: SnRefContext) -> None:
+        diag_layer = odxrequire(context.diag_layer)
+
+        if self.diag_comm_snref is not None:
+            self._diag_comm = resolve_snref(self.diag_comm_snref, diag_layer.diag_comms, DiagComm)
+
+        service = self.diag_comm
+        if not isinstance(service, DiagService):
+            odxraise(f"DIAG-VARIABLE references non-service {type(service).__name__} "
+                     f"diagnostic communication")
+
+        self._in_param_if = None
+        if self.in_param_if_snref is not None:
+            self._in_param_if = resolve_snref(self.in_param_if_snref,
+                                              odxrequire(service.request).parameters, Parameter)
+        else:
+            odxraise(f"No IN-PARAM-IF referenced (note that SNPATHREFs are not yet supported)")
+
+        self._out_param_if = None
+        if self.out_param_if_snref is not None:
+            self._out_param_if = resolve_snref(self.out_param_if_snref,
+                                               odxrequire(service.positive_responses[0]).parameters,
+                                               Parameter)
+        else:
+            odxraise(f"No OUT-PARAM-IF referenced (note that SNPATHREFs are not yet supported)")
```

Comment:
This does not make the situation any better, odxraise should only and only be used when the ODX file is wrong

Reply:
yeah, that was my previous point: we do not implement SNPATHREFs (yet?) and thus we only have the following options:

1) do not resolve the param objects at all. this would make it pretty hard to actually use the functionality and would be contradictory to basically any other place where references are involved. 
2) make it work for the approximately 95% of the cases where SNREFs are used and
  (a) complain about SNPATHREFs during load time when in strict mode
  (b) raise an (almost certainly unexpected) exception once the object referenced via SNPATHREF is to be accessed by the user code.

originally, this PR was in camp (2a), currently it is using (2b). I think (2b) is not as bad as it looks, though, because I'm pretty sure that 99% of all odxtools users will never touch the diagnostic variables functionality. That said, maybe I missed something...

Reply:
ok, I guess I misunderstood what you meant: you said even with the latest version of the PR we were essentially still in camp (2a)? I removed these `odxraise` statements, so we are hopefully in (2b) now...

PS: do you have files which define diag variables? if yes, it would be cool if you could test if this code works...